src/extensions/extra/selfie-taker/camera.js

```diff
@@ -0,0 +1,201 @@
+            height = width / (4/3);
+        }
+
+        video.setAttribute('width', width);
+        video.setAttribute('height', height);
+        canvas.setAttribute('width', width);
+        canvas.setAttribute('height', height);
+        streaming = true;
+    }
+
+    function initWidget() {
+        var selfieContainer = $("#selfie-container");
+        $("#selfie-allow-access").remove();
+        snapAudio = new Audio("./extensions/default/ImageUrlCodeHints/camera-shutter-click-08.mp3");
+        selfieContainer.prepend(selfieWidgetHTML);
+        video = document.getElementById("selfie-video");
+        photo = document.getElementById("selfie-photo");
+        snap = document.getElementById("selfie-snap");
+        canvas = document.getElementById("selfie-canvas");
+        use = document.getElementById("selfie-use");
+        use.style.display = "initial";
+    }
+
+    function playVideo(stream) {
+        initWidget();
+
+        video.src = vendorURL.createObjectURL(stream);
+
+        video.addEventListener("canplay", streamVideo, false);
+        video.addEventListener("pause", function() {
+            vendorURL.revokeObjectURL(video.src);
+
+            if(stream) {
+                stream.stop();
+                stream = null;
+                streaming = false;
+            }
+        });
+        snap.addEventListener("click", function(ev) {
+            ev.preventDefault();
+            use.removeAttribute("disabled");
+
+            $("#selfie-video-bg").addClass("on");
+
+
+            snapAudio.play();
+            setTimeout(function() {
+                $("#selfie-video-bg").removeClass("on");
+            }, 105);
+
+            snapPhoto(deferred, filePath);
+        });
+
+        video.play();
+    }
+
+    function snapPhoto() {
+        var context = canvas.getContext("2d");
+        var data;
+
+        function savePhoto() {
+            if(!data) {
+                return deferred.reject();
+            }
+
+            var bin = /^data:image\/png;base64,(.+)/.exec(data)[1];
+            save(deferred, filePath, base64ToBuffer(bin));
+        }
+
+        if(!width || !height) {
+            return deferred.reject();
+        }
+
+        canvas.width = width;
+        canvas.height = height;
+        context.drawImage(video, 0, 0, width, height);
+
+        // TODO: Revoke data url on close
+        data = canvas.toDataURL("image/png");
+        photo.setAttribute("src", data);
+        use.removeEventListener("click", savePhoto);
+        use.addEventListener("click", savePhoto);
+    }
+
+    function attachCamera() {
+        if(navigator.getMedia) { // Standard
+            navigator.getMedia({video: true, audio: false}, playVideo, function(error) {
+                deferred.reject(error);
+            });
+        }
+    }
+
+    function getSelfieDialog() {
+        return _selfieDialog;
+    }
+
+    function show(path) {
+        deferred = new $.Deferred();
+        filePath = path;
```

Comment:
I'd move your other functions above that need access to these variables inside this function, so they can reach them via the scope chain vs. globals.